### ISSUE TITLE
Double check the missing `LN `in `IGAB`

### ISSUE THREAD

**Issue by zzowenzz** [caiyuanhao1998/Retinexformer#209]
Hi @caiyuanhao1998 , regarding of the missing LN in the architecture figure, I understand that the implementation is inside the IG-MSA with `q = F.normalize(q, dim=-1, p=2)` and `k = F.normalize(k, dim=-1, p=2)` in the forward function of that class, which is also reflected from issue #13 and #89. My question is:

1. Why you only implement ` F.normalize` on Q and K, instead of V? Any ablation between with and without on V?
2. ` F.normalize` is not strictly LayerNorm as it performs more like cosine similarity when apply the dot product between Q and K, as $Q^{\prime} = \frac{Q}{||Q||}$ and $K^{\prime} = \frac{K}{||K||}$. Then $Q^\prime \cdot K^\prime = \frac{Q\cdot K}{||Q||||K||}$. which is just cosine similarity. So did you compare the empirical difference between using ` F.normalize` and `nn.LayerNorm`? And of course using ` F.normalize` makes less learnable parameter. 
3. Since using ` F.normalize`  is more like cosine similarity between Q and K, which makes them missing the magnitude information, from tensor perspective. Does it makes the designed IG-MSA module have less information to learn as a result?

**Comment by caiyuanhao1998** [comment]
Hi, great questions — let me clarify the design choices.

### Why L2-normalize **Q/K** but not **V**?
- **Effect on logits.** `F.normalize(q)` and `F.normalize(k)` make the attention score proportional to **cosine similarity**:  
$\text{logits}_{ij} \propto \frac{q_i^\top k_j}{\|q_i\|\|k_j\|}$
  This removes scale variation in token features and stabilizes training (especially under mixed precision), while keeping a single global scale via the temperature (the usual `1/sqrt(d)` or a learnable scalar).
- **Why not normalize V.** The **values** carry the content magnitude that the model uses downstream (residual/MLP). Normalizing `V` would erase that amplitude information and empirically tends to reduce representational capacity (the attention weights already discard magnitude by using cosine on Q/K; we don’t want to also flatten the value stream).

### “`F.normalize` vs `nn.LayerNorm`”
They serve different purposes:
- **`F.normalize` (L2)**: direction-only, no learnable affine, applied to Q/K; turns dot-product into cosine similarity + global temperature. Invariant to per-token scale.
- **`nn.LayerNorm`**: mean/variance normalization **with learnable affine** on each token. It changes the feature distribution (shift/scale) but does **not** specifically enforce cosine geometry.

Because of this, swapping `F.normalize` for `LayerNorm` on Q/K is not apples to apples: you’d change the geometry **and** add learnable scales/biases. If you try it, you typically need to **retune the temperature** (and sometimes dropout/LR).

### Ablations
- We did not find a clear benefit from normalizing **V**, so the released code only normalizes Q/K.
- If you’d like to compare, here is a minimal flag you can add to IG-MSA to toggle the behavior:

```python
# inside forward of IG-MSA
use_cosine = True          # cosine attention on Q/K
normalize_v = False        # keep V unnormalized by default
learned_tau = getattr(self, "tau", None)  # optional learned temperature

q = self.q_proj(x)
k = self.k_proj(x)
v = self.v_proj(x)

if use_cosine:
    q = F.normalize(q, dim=-1, p=2)
    k = F.normalize(k, dim=-1, p=2)
    scale = (learned_tau if learned_tau is not None else 1.0 / math.sqrt(q.size(-1)))
else:
    # e.g., LayerNorm on Q/K (not equivalent to cosine)
    q = self.q_ln(q)  # nn.LayerNorm(d_model) defined in __init__
    k = self.k_ln(k)
    scale = (learned_tau if learned_tau is not None else 1.0 / math.sqrt(q.size(-1)))

if normalize_v:
    v = F.normalize(v, dim=-1, p=2)  # usually hurts; off by default

attn = torch.softmax(torch.matmul(q, k.transpose(-2, -1)) * scale, dim=-1)
y = torch.matmul(attn, v)
```

Here are the tips for a fair comparison:

1. If you switch off cosine (no L2 on Q/K) or replace with LayerNorm, re-tune the temperature (τ) — cosine often prefers a slightly higher τ.

2. Keep the rest of the stack identical (same dropout/LR/warmup) and report the same metrics.



If you find our repo useful, please help us star it, thanks :)
